Add support for namespaces in proxy registry

DerEnderKeks opened the pull request:
This PR adds support for multiple remote registries when using distribution in proxy mode using the `ns` query parameter, which is proposed to the OCI spec [here](https://github.com/opencontainers/distribution-spec/pull/66) and already implemented in Containerd.

I implemented it in a mostly backwards compatible way, the old proxy config settings remain intact. When the new config options are not used, it behaves as before. When namespaces are enabled, all remote repositories are stored prepended with the domain of the remote. The remote is either determined using the `ns` query parameter, or when missing using the domain that was prepended to the image name.

Pulling an image the following ways would both pull the image `some/image:latest` from `external.example.org`:

- Pulling `registry-proxy.example.com/some/image:latest`, with `ns` set to `external.example.org` 
- Pulling `registry-proxy.example.com/external.example.org/some/image:latest`

To enable the feature one would put the following in the config file:

```yaml
proxy:
  enablenamespaces: true
  credentials: # optional
    'https://example.com':
      username: someuser
      password:  somepassword
```

I'd appreciate any feedback.

milosgajdos commented:
You need to sign DCO as per the contributor's guidelines.

dcarrion87 commented:
@DerEnderKeks this looks useful!

We're wanting to support multiple registries and just wondering if this would solve our problem. We run containerd like this:

```
[plugins.cri.registry]
  [plugins.cri.registry.mirrors]
    [plugins.cri.registry.mirrors."*"]
      endpoint = ["https://myregistry.mydomain.com"]
```

But as you know if say `ghcr.io/myimage` comes through it bypasses the mirror.

Just looking at the code it just seems like as long as `ns=` comes through and `enablenamespaces: true` this should just work?

dcarrion87 commented:
Well that _just worked_. This feature is great! Would love to see this merged @milosgajdos @DerEnderKeks!

dcarrion87 commented:
Just bumping this to see what is necessary to have this merged? This solves a few issues that have been raised in this project for multi remote.

marco-svitol commented:
Would love to have this feature! We are using proxy registry to speed up our Kind cluster integration tests. As now we must create one proxy registry per remote registry.
This feature would enale us to use a single proxy.

dcarrion87 commented:
@marco-svitol we ended up forking and using our own image as it's crickets on when this is going to be merged. It's been running for a week and so far so good. We've got some hefty images across dockerhub and ghcr going through it.

marco-svitol commented:
@dcarrion87 is the image publicly available? I would like to test it too.
thanks!

milosgajdos commented:
This is building support for an OCI _proposal_ that has not been merged. I'd encourage you to ask OCI maintainers on the OCI PR proposal to consider merging it. We are not against adding this in.

dcarrion87 commented:
I bumped https://github.com/opencontainers/distribution-spec/pull/66

hoozecn commented:
seems not working for upstream with anonymous credential

DerEnderKeks commented:
Works fine for me. I tested with docker.io, quay.io and probably some others. 
You got to be a little more specific in what you mean with "not working". Any related log output? What exactly doesn't work?

hoozecn commented:
> Works fine for me. I tested with docker.io, quay.io and probably some others. You got to be a little more specific in what you mean with "not working". Any related log output? What exactly doesn't work?

Here was how I tested:

1. git clone the latest version of https://github.com/distribution/distribution
2. merge this PR
3. run`go run ./cmd/registry serve config.yaml`

config.yaml
```
version: 0.1
log:
  accesslog:
    disabled: false
  level: debug
  formatter: json
loglevel: debug
storage:
  filesystem:
    rootdirectory: /tmp/registry
    maxthreads: 100
http:
  addr: localhost:5000
proxy:
  enablenamespaces: true
```

4. run `docker pull localhost:5000/docker.io/library/busybox:latest`

got an error response:
```
Error response from daemon: manifest for localhost:5000/docker.io/library/busybox:latest not found: manifest unknown: manifest unknown
```

hoozecn commented:
@DerEnderKeks I tried to fix the problem and it seems to work with the patch 

https://github.com/DerEnderKeks/distribution/pull/1/files

DerEnderKeks commented:
The issue only occurred when pulling from docker.io *and* with the domain in the path, not when using the ns parameter, because I previously extracted the remote URL after setting up the credentials. This is now fixed.

@hoozecn your PR pointed me to the right place, thanks.

dtomasi commented:
Is there any progress? This feature would be extremely useful for me.

Jamstah commented:
@DerEnderKeks Can you please rebase this?

One quick review comment, there appear to be no docs changes here, which there should be with configuration format changes. When that's done, can we please indicate in some way that the feature is based on a spec proposal and subject to change?

darzanebor commented:
greate feature, any updates when it would be merged?

DerEnderKeks commented:
I rebased my branch and added documentation for the new config options. As requested, I specificly mentioned that this feature is based on a proposal.

Is there anything else required to get this merged?

darzanebor commented:
@DerEnderKeks could you please fix issue "Clear-text logging of sensitive information"? I think it's the final issue which blocks this feature.

DerEnderKeks commented:
I don't quite understand why CodeQL thinks that the password is logged. The password is never passed to any logging call and the check only failed in places I didn't even modify. Maybe a false positive?

darzanebor commented:
@milosgajdos could you please advise is it really false positive in job CodeQL?

milosgajdos commented:
We've had issues with CodeQL acting silly in the past and when the GH team introduced some bugs. Not quite sure what's going on here tbqh.

darzanebor commented:
@marco-svitol @Jamstah could you please merge it?

milosgajdos commented:
> @marco-svitol @Jamstah could you please merge it?

Before anything is merged we'll need the referenced distribution-spec proposal to be merged (https://github.com/opencontainers/distribution-spec/pull/66)

marco-svitol commented:
> @marco-svitol @Jamstah could you please merge it?

I'm not in charge of this nor I have the authorization.

S0ly commented:
any update on that ?

dkrhodes commented:
It's a great feature. When do you think it can be merged? @milosgajdos
